dbms/src/DataStreams/LiveViewBlockInputStream.h

```diff
@@ -0,0 +1,222 @@
+        storage->condition.broadcast();
+    }
+
+    Block getHeader() const override { return storage->getHeader(); }
+
+    void refresh()
+    {
+        if (active && blocks && it == end)
+            it = blocks->begin();
+    }
+
+    void suspend()
+    {
+        active.reset();
+    }
+
+    void resume()
+    {
+        active = active_ptr.lock();
+        {
+            if (!blocks || blocks.get() != (*blocks_ptr).get())
+                blocks = (*blocks_ptr);
+        }
+        it = blocks->begin();
+        begin = blocks->begin();
+        end = blocks->end();
+    }
+
+    NonBlockingResult tryRead()
+    {
+        return tryRead_(false);
+    }
+
+protected:
+    Block readImpl() override
+    {
+        /// try reading
+        return tryRead_(true).first;
+    }
+
+    /** tryRead method attempts to read a block in either blocking
+     *  or non-blocking mode. If blocking is set to false
+     *  then method return empty block with flag set to false
+     *  to indicate that method would block to get the next block.
+     */
+    NonBlockingResult tryRead_(bool blocking)
+    {
+        Block res;
+
+        if (has_limit && num_updates == static_cast<Int64>(limit))
+        {
+            return { Block(), true };
+        }
+        /// If blocks were never assigned get blocks
+        if (!blocks)
+        {
+            Poco::FastMutex::ScopedLock lock(storage->mutex);
+            if (!active)
+                return { Block(), false };
+            blocks = (*blocks_ptr);
+            it = blocks->begin();
+            begin = blocks->begin();
+            end = blocks->end();
+        }
+
+        if (isCancelled() || storage->is_dropped)
+        {
+            return { Block(), true };
+        }
+
+        if (it == end)
+        {
+            {
+                Poco::FastMutex::ScopedLock lock(storage->mutex);
+                if (!active)
+                    return { Block(), false };
+                /// If we are done iterating over our blocks
+                /// and there are new blocks availble then get them
+                if (blocks.get() != (*blocks_ptr).get())
+                {
+                    blocks = (*blocks_ptr);
+                    it = blocks->begin();
+                    begin = blocks->begin();
+                    end = blocks->end();
+                }
+                /// No new blocks available wait for new ones
+                else
+                {
+                    if (!blocking)
+                    {
+                        return { Block(), false };
+                    }
+                    if (!end_of_blocks)
+                    {
+                        end_of_blocks = true;
+                        return { getHeader(), true };
+                    }
+                    while (true)
+                    {
+                        UInt64 timestamp_usec = static_cast<UInt64>(timestamp.epochMicroseconds());
```

Comment:
`Timestamp` stores value that never changes. `timestamp.epochMicroseconds() ` is implemented as:
```c++
inline Timestamp::TimeVal Timestamp::epochMicroseconds() const
{
	return _ts;
}
```
You can use default constructor `Timestamp().epochMicroseconds()` and don't store `timestamp` field.

Reply:
Or call `timestamp.update()` each time.

Reply:
Oh, also it's even not correct to use `Timestamp` class here.
Because `Timestamp::update()` is implemented via `system_clock`:
```c++
void Timestamp::update()
{
	_ts = std::chrono::duration_cast<std::chrono::microseconds>
		(std::chrono::system_clock::now().time_since_epoch()).count();
}
```
Actually, you can copy this code and replace `system_clock` to `steady_clock`.
See https://stackoverflow.com/questions/13263277/difference-between-stdsystem-clock-and-stdsteady-clock